persist/fs/index_lookup.go

```diff
@@ -0,0 +1,197 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+
+	"github.com/m3db/m3db/digest"
+	"github.com/m3db/m3db/persist/encoding"
+	"github.com/m3db/m3db/ts"
+
+	"gopkg.in/vmihailenco/msgpack.v2"
+)
+
+// nearestIndexOffsetLookup provides a way of quickly determining the nearest offset of an
+// ID in the index file. It is not safe for concurrent use
+type nearestIndexOffsetLookup struct {
+	summaryIDsOffsets []encoding.IndexSummaryToken
+	// bytes from file mmap'd into anonymous region
+	summariesBytes []byte
+	// reusable decoder stream
+	decoderStream  encoding.DecoderStream
+	decoder        encoding.Decoder
+	msgpackDecoder *msgpack.Decoder
+}
+
+func newNearestIndexOffsetLookup(
+	summaryIDsOffsets []encoding.IndexSummaryToken,
+	summariesBytes []byte,
+	decoder encoding.Decoder,
+	decoderStream encoding.DecoderStream,
+) *nearestIndexOffsetLookup {
+	return &nearestIndexOffsetLookup{
+		summaryIDsOffsets: summaryIDsOffsets,
+		summariesBytes:    summariesBytes,
+		decoderStream:     decoderStream,
+		decoder:           decoder,
+		msgpackDecoder:    msgpack.NewDecoder(nil),
+	}
+}
+
+// getNearestIndexFileOffset returns either:
+//     1. The offset in the index file for the specified series
+//     2. The offset in the index file for the the series in the summaries file
+//        that satisfies the following two constraints:
+//            1. Is closest to the desired series in the index file
+//            2. Is BEFORE the desired series in the index file (because we
+//               we scan the index file sequentially in a forward-moving manner)
+// In other words, the returned offset can always be used as a starting point to
+// begin scanning the index file for the desired series.
+func (il *nearestIndexOffsetLookup) getNearestIndexFileOffset(id ts.ID) (int64, bool, error) {
+	idBytes := id.Data().Get()
+
+	min := 0
+	max := len(il.summaryIDsOffsets) - 1
+
+	// The summaries file only contains a fraction of the series that are in
+	// the index file itself. Because of that, the binary search that we're
+	// performing is "optimistic". We're trying to find either an exact match,
+	// OR the nearest match that is to the left of the series we're searching
+	// for (so we keep track of it everytime we move right).
+	bestMatchSoFar := int64(-1)
+
+	for {
+		if min > max {
+			if bestMatchSoFar == -1 {
+				return -1, false, nil
```

Comment:
Hm, this is dependent on the first index entry always being available as a summary yeah?  I wonder if we should depend on this or not and just start searching from the start of the index if there's no match?

Reply:
You're right, it would be much better not to depend on that being true. If thats the case, then I can probably get rid of the bool in the return value ("found"), and just have it return 0 cause this is the only case where it can happen.